Improve and unify developer shell interfaces

psakievich opened the pull request:
## Updated Scope:

This PR adds functionality to `spack dev-build` and `spack build-env` to increase compatibility and add a longstanding external feature `build-env-dive` to spack core.

`build-env-dive [spec]` is a shell function that 

1. Launches a subshell with the build-environment for the spec active
2. Navigates to the build directory of the spec
3. Changes the prompt to indicate to the user they are in the build-env subshell

The features added are:
- `spack build-env`
   - add `--dive` for diving into the build-env 
   - change the prompt as part of the dive process (bourne shells and csh only for now)
   - add `spack build-env --status` for people using shells that don't have a changed  prompt (aka fish, powershell)
   - add `--cd` for navigation with subshell commands, (reproduces `spack cd` as a precursor to the dive/cmd)
- `spack dev-build`
   - allow it to work with spack environments (requires specs to be marked as develop)
   - remove constraint for concrete spec
   - add `--prompt` and unify code for the `--drop-in` with `spack build-env --dive`
- aliases
   - create a `spack dev-dive` alias to replace `build-env-dive` completely   

### Example developer workflow   
``` shell
$ spack env activate --temp
$ spack add zlib-ng
$ spack stage -p $SPACK_ENV zlib-ng
$ spack develop zlib-ng
$ spack install -u autoreconf zlib-ng
# fails
$ spack dev-dive zlib-ng
# prompt changed, let's check the location
zlib-ng-build-env $ pwd
/private/var/folders/ln/1_3kxbwd35s_ylsjlm3zmqmc00307v/T/spack-ne8_m488/zlib-ng
# confirm build-env in case we forget where we are in subshell hierarchy
zlib-ng-build-env $ spack build-env --status
==> In build env zlib-ng-wrsaadvkbg7rjj7kfjw5rhdrrfdswcmm
# fix code
zlib-ng-build-env $ make -j6
# builds  now
zlib-ng-build-env $ exit
$ spack build-env --status
==> build environment not detected
$ spack install
```

## Original message:

There is really no reason why `spack develop` and `dev-build` shouldn't play along with each other.  `dev-build` just assumes it can't work in an environment. This PR makes it so `dev-build` will look for matching develop specs in a concrete environment.

We could optionally make it do a `spack add && spack develop` if they don't exist but I think that is too far reaching.

We have constantly been re-writing a build-env-dive shell function that

1. loads the build environment for a develop spec in a subshell
5. spack cd -b to the build directory

Item 1 is already in place with `spack dev-build` so rather than write a whole new funciton, I think we should make them compatible.

That said I think the `dev-build` interface is a little clunky for how `build-env-dive` has been used and I think it needs to be improved, but it might just be doable through aliases. 

Things I want to add still:

- [x] prompt to indicate you are in the build environment when using --build-dive
- [x] navigation option, dive then go to build or src directory

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

psakievich committed:
Make dev-build compatible with spack develop

haampie commented on lib/spack/spack/cmd/dev_build.py:
```diff
@@ -102,29 +103,40 @@ def dev_build(self, args):
+            tty.msg("Uninstall or try adding a version suffix for this dev build.")
+            sys.exit(1)
+
+        # disable checksumming if requested
+        if args.no_checksum:
+            spack.config.set("config:checksum", False, scope="command_line")
```
This can be deleted generally, right?

tgamblin replied:
There's no checksum arg action (though maybe there should be), so currently commands have to do this.

haampie replied:
point is no fetching happens because dev_path is set

tgamblin replied:
I believe it may fetch patches (unless you `--skip-patch`) -- not remembering if it fetches resources separately too but pretty sure it does b/c we don't have special handling for that.

haampie replied:
Okay, then it should not be indented. Move before the new branch?

psakievich replied:
> This can be deleted generally, right?

I think so?  TBH I haven't used this feature at all since `spack develop` became a thing.  It basically does yet another wrapper around `spack install` which I'm not a huge fan of.  I started working on this because there are some SNL people who mentioned they use it and it bugs me that it can't be used with environments. 

My only concern with deleting this is for those who use this without an environment (which is the entire user base) it may break them. But really we shouldn't use this flag anyways. I think this whole command needs to be revamped.

psakievich replied:
Hmm missed all the discussion. Didn't update browser. This DOES fetching.  It works without an environment and basically let's you develop a single root spec. So these fetch commands also apply to the dependencies IIRC.  I've just been trying to get it to work with environments.

haampie commented:
`spack dev-build --help` and docs should be updated.

As for going to the build environment, what's wrong with `spack build-env`? That it's lacking `--drop-in`? Adding that flag sounds useful and trivial

tgamblin commented on lib/spack/spack/cmd/dev_build.py:
```diff
@@ -102,29 +103,40 @@ def dev_build(self, args):
+            tty.die(
+                "spack dev-build spec must have a single, concrete version. "
+                "Did you forget a package version number?"
+            )
```
I think this should match #46911

psakievich commented:
> `spack dev-build --help` and docs should be updated.
> 
> As for a shell with the build environment, what's wrong with `spack build-env`? That it's lacking `--drop-in` and doesn't change the working dir? Adding that flag sounds useful and trivial.
> 
> I've never used `spack dev-build --drop-in`, looks very weird to me that it happens past `.install()` and won't happen on install failure. Do you agree?
> 
> To me it would make more sense if:
> 
> * `spack build-env --drop-in` would do what it says without install
> * `spack install/dev-build --drop-in` would drop in on install failure

Yes I agree. Adding a `--drop-in`, I would prefer `--dive`, for `build-env` makes a lot of sense.  The drop-in still happens even if the install fails for this command.  The `PackageInstaller.install()` can fail but the command continues to the `drop-in` section.  Overall I find this pretty unintuitive.

I would propose the following for `dev-build`:

Stage 1. Make it compatible with environments and `spack develop` and advertise to the user base who uses it.
Stage 2. Determine how much use it gets
Stage 3. Deprecate and then major refactor or remove

Overall I think the `dev-build` may command have a place, but when I'm developing I don't want to run a command with a ton of options since development CLI commands tend to be repetitive and frequent.

I don't like that it is essentially wrapping `spack install` for non-environments and I find the current interface clunky.  However for quick and dirty develop it seems like it could be useful if it is refactored and updated.

psakievich committed:
Fix some things from review

psakievich committed:
Update help

psakievich commented:
I think I might give up on the prompt for this PR. I think adding that plus a `--drop-in/--dive` option to `build-env` might be suitable for a follow on PR. I think I'll put the navigation option in either with or after that.  Really I'd like to give more thought to cleaning up the relationship between these two. I spent a couple hours today trying to get the prompt stuff to work, but I don't think it is necessary for this PR.

I personally would classify this as a bug fix since there was no real reason for this command to not work with environments in the past.

@haampie I've updated the help section. This command does not appear in the docs 😬 .  Not sure I want to add it right now given I consider it a candiadate to be reworked or deleted.

psakievich committed:
Add prompt capability to dev_build drop-in

psakievich committed:
Merge remote-tracking branch 'origin' into psakiev/dev-build

psakievich committed:
WIP unify

psakievich committed:
Add build-env --dive

psakievich committed:
Merge remote-tracking branch 'origin' into psakiev/dev-build

psakievich committed:
Add completion

psakievich committed:
Fixes

psakievich committed:
Compatibility

psakievich committed:
Add navigation to dive feature

psakievich committed:
Style

psakievich committed:
Add docs

psakievich committed:
Re-write test and update docs some more

psakievich committed:
Add a cd test

psakievich committed:
[@spackbot] updating style on behalf of psakievich

psakievich commented on lib/spack/spack/prompt.py:
```diff
@@ -0,0 +1,73 @@
+    else:
+        bash_color_prompt = colorize(f"@G{{{prompt}}}", color=True, enclose=True)
+        zsh_color_prompt = colorize(f"@G{{{prompt}}}", color=True, enclose=False, zsh=True)
+        cmds += textwrap.dedent(
+            rf"""
+            if [ -z ${{SPACK_OLD_PS1+x}} ]; then
+                if [ -z ${{PS1+x}} ]; then
+                    PS1='$$$$';
+                fi;
+                export SPACK_OLD_PS1="${{PS1}}";
+            fi;
+            if [ -n "${{TERM:-}}" ] && [ "${{TERM#*color}}" != "${{TERM}}" ] && \
+               [ -n "${{BASH:-}}" ];
+            then
+                export PS1="{bash_color_prompt} ${{PS1}}";
+            elif [ -n "${{TERM:-}}" ] && [ "${{TERM#*color}}" != "${{TERM}}" ] && \
+                 [ -n "${{ZSH_NAME:-}}" ];
+            then
+                export PS1="{zsh_color_prompt} ${{PS1}}";
+            else
+                export PS1="{prompt} ${{PS1}}";
+            fi
+            """
+        ).lstrip("\n")
```
Is there any reason why we do this logic as shell conditionals rather than just inspecting the environment like I do in the function above? Seems like all of this could just get deleted and go into the environment mod. The `fish` prompt change doesn't even do anything at the moment...

tgamblin replied:
Two things:

1. You're not guaranteed that the environment in `spack` is the same environment as that of the user's shell. I could see people putting wrappers around `spack` or other weird things that would make this not work the way you might think it should.
2. This logic has to save state in the user's shell environment so that we can put things back the way they were after deactivation.  And the saved state (the prompt) needs to come from the user's shell, not `spack`'s environment.

tgamblin replied:
Here's an illustrative example -- the shell detection stuff is generally not available from the environment of things launched from that shell:

```console
(spackbook):~$ echo $BASH
/bin/bash
(spackbook):~$ python -c 'import os; os.environ["BASH"]'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<frozen os>", line 679, in __getitem__
KeyError: 'BASH'
```

```console
(spackbook):~> echo $ZSH_NAME
zsh
(spackbook):~> python -c 'import os; os.environ["ZSH_NAME"]'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<frozen os>", line 679, in __getitem__
KeyError: 'ZSH_NAME'
```

psakievich replied:
Okay yeah I noticed this in that the current prompt is not getting preserved with the dive. So how do we decide what goes into an `EnvironmentModification` object vs what goes into the shell wrapper?  Trial and error?

psakievich replied:
The save state (bullet 2) seems like it would apply either way since we are dumping all of the env mods to a shell wrapper anyways...

tgamblin replied:
> So how do we decide what goes into an EnvironmentModification object vs what goes into the shell wrapper? Trial and error?

if you need to read directly from some variable in the user's environment that's *only* in the shell, you have to code things up like this.  This is not most things (e.g. PATH, etc. are preserved).  There are some nasty interactions to be aware of, e.g. macOS clears out some variables (DYLD_LIBRARY_PATH) on process launch -- see setup-env.sh for more on that) and Spack sanitizes the build environment--not sure that is relevant here but see #46823.

Geekdude committed:
Add prompt support for fish. (#47609)

To avoid changing the fish_prompt function, create a SPACK_PROMPT
environment variable, which users can then manually incorporate into
their fish_prompt.fish function.

psakievich commented on lib/spack/spack/prompt.py:
```diff
@@ -0,0 +1,74 @@
+from spack.util.environment import EnvironmentModifications
+
+
+def prompt_modifications(prompt, shell, env=os.environ):
```
@tgamblin based on your comments I take it this is a bad idea. Do you have any thoughts on how we could make this work?  I could a also just abandon preserving the users prompt for the `--dive`.  I'm okay with that personally. I mainly want a clear indicator you switched to the subshell.

psakievich committed:
Style

psakievich committed:
Revert prompt changes for environment activation

I tried consolidating but I am not convinced it will work
correctly

psakievich committed:
Update imports

johnwparent commented on lib/spack/spack/test/cmd/build_env.py:
```diff
@@ -58,6 +63,31 @@ def test_pickle(tmpdir):
+@pytest.mark.usefixtures("config", "mock_packages", "working_env")
+def test_cd(cd_key, tmpdir, monkeypatch, capfd):
+    """test that a subshell will navigate using spack cd before running commands"""
+    cmd = "pwd" if sys.platform != "win32" else "Get-Location"
```
This line is causing the current CI failure

On posix systems `pwd` is a binary (or at least an executable script) available in the PATH. However on Windows, `Get-Location` is a powershell commandlet, meaning it has no meaning to the system if invoked via `exec` (or anything that isn't powershell). 

There's no good Windows equivalent of the `pwd` binary, you can either do some fancy work with direct win32 API calls, or just invoke a shell when executing the command (beware if it's powershell you'll need to post-process the value returned by `Get-Location` as it will be a powershell object). That could look something like:
```suggestion
    cmd = "pwd" if sys.platform != "win32" else 'powershell.exe -Command "& {(Get-Location).Path}'
```

johnwparent commented:
Left a review comment, still need to take a closer pass for Windows concerns, but this should at least unblock the CI.

psakievich committed:
Update build_env.py

Co-authored-by: John W. Parent <45471568+johnwparent@users.noreply.github.com>

psakievich commented on lib/spack/spack/test/cmd/build_env.py:
```diff
@@ -58,6 +63,31 @@ def test_pickle(tmpdir):
+@pytest.mark.usefixtures("config", "mock_packages", "working_env")
+def test_cd(cd_key, tmpdir, monkeypatch, capfd):
+    """test that a subshell will navigate using spack cd before running commands"""
+    cmd = "pwd" if sys.platform != "win32" else 'powershell.exe -Command "& {(Get-Location).Path}'
```
```suggestion
    cmd = "pwd" if sys.platform != "win32" else 'powershell.exe -Command "& {(Get-Location).Path}"'
```

psakievich committed:
Update lib/spack/spack/test/cmd/build_env.py

johnwparent commented on lib/spack/spack/test/cmd/build_env.py:
```diff
@@ -58,6 +63,31 @@ def test_pickle(tmpdir):
+
+        pwd = os.getcwd()
+        spec = Spec("zlib").concretized()
+        run_command_in_subshell(spec, Context.BUILD, [cmd], cd_arg=cd_key)
```
```suggestion
        run_command_in_subshell(spec, Context.BUILD, cmd, cd_arg=cd_key)
```

johnwparent replied:
`subprocess.check_output` takes either a list of command + args or a string with the whole invocation, composing the Windows command line into a list causes the entire powershell invocation to be interpreted as a single command, which of course cannot be found.

johnwparent commented on lib/spack/spack/util/shell_detection.py:
```diff
@@ -0,0 +1,35 @@
+                    pass
+            except subprocess.CalledProcessError:
+                pass
+            raise SpackError("Unknown shell type being used on Windows")
```
Do you want me to add support for CMD?

psakievich replied:
Hmm what do you think is best? I could just punt on windows for this PR and you could do a follow on? Otherwise likely yes? I never use windows so I don't really know what is best here.

johnwparent replied:
I can also just commit directly to this PR if you're cool with that, shouldn't take long

psakievich replied:
Yeah that is fine if the changes are small. If I run into too many more windows issues then I think we should break it into two PR's since I can't really test windows behavior outside of CI at the moment.

johnwparent replied:
Agreed, that's reasonable.

johnwparent commented on lib/spack/spack/util/shell_detection.py:
```diff
@@ -0,0 +1,35 @@
+        else:
+            try:
+                output = subprocess.check_output(
+                    'powershell -Command "echo $PSVersionTable"', universal_newlines=True
```
This will always pass (fine if we don't want to support CMD), as the powershell executable is always in the path unless a user removed it explicitly. 

I think something like:
```
pproc_name = psutil.Process(os.getppid()).name()
is_power_shell = bool(re.fullmatch('pwsh|pwsh.exe|powershell|powershell.exe', pproc_name))
```

With a better regex would be best.

johnwparent commented on lib/spack/spack/prompt.py:
```diff
@@ -0,0 +1,74 @@
+    mods = EnvironmentModifications()
+
+    if shell == "fish" or shell == "pwsh" or shell == "bat":
+        # requires a function and can't be set with os.environ
```
bat can be set from env variables

johnwparent reviewed (commented):
Partial review, still need to look closer at the dev_build and location commands.

Trying this on Windows results in an error, tracing that down now.

psakievich committed:
Update lib/spack/spack/test/cmd/build_env.py

Co-authored-by: John W. Parent <45471568+johnwparent@users.noreply.github.com>

psakievich committed:
Merge branch 'develop' into psakiev/dev-build

psakievich committed:
Style

psakievich committed:
Fix unit test

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Delete stale doc

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Style

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
License

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Merge branch 'develop' into psakiev/dev-build

Signed-off-by: Phil Sakievich <psakiev@sandia.gov>

psakievich committed:
Delete prompt mods for now

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Fix failing test

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Fix more tests

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Refactor more tests

Signed-off-by: psakiev <psakiev@sandia.gov>
